Populate sensitivities for single result view

dhruvkb opened the pull request:
<!-- prettier-ignore -->
## Fixes
<!-- If PR doesn't fully resolve the issue, replace 'Fixes' below with 'Related to'. -->
<!-- If there is no issue being resolved, please consider opening one before creating this pull request. -->

Fixes #2926 by @sarayourfriend 

## Description
<!-- Concisely describe what the pull request does. -->
<!-- Add screenshots, videos, or other media to show the problem and the solution when appropriate. -->

This PR populates the sensitivies for single result view by manually defining the `retrieve` view in the `MediaViewSet` and extending the `SearchContext` to work for single results too.

## Testing Instructions
<!-- Give steps for the reviewer to verify that this PR fixes the problem; or delete this section entirely. -->

1. Check out the branch and run the dev server `just api/up` (and `just api/init` if you haven't already)
0. Find a media item that has some sensitive content. To do so, add the the query param `unstable__include_sensitive_results=true` to the URL.
1. Open the same media's detail page, you should see the same sensitivities as the results page.
2. Report some media item as mature. Do the same for this media item. You should see identical sensitivity in the search resuls and single result pages.

## Checklist
<!-- Replace  the [ ] with [x] to check the boxes. -->

- [x] My pull request has a descriptive title (not a vague title like`Update index.md`).
- [x] My pull request targets the _default_ branch of the repository (`main`) or a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [ ] I added or updated tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.
- [ ] I ran the DAG documentation generator (if applicable).

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin
<!-- You must read and understand the following attestation. -->

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

sarayourfriend commented on api/api/utils/search_context.py:
```diff
@@ -41,7 +44,12 @@ def build(cls, results: list[Hit], origin_index: OriginIndex) -> Self:
             # cf: https://github.com/WordPress/openverse/issues/2154
             Q(
                 "terms",
-                **{"identifier.keyword": [result.identifier for result in results]},
+                **{"identifier.keyword": all_result_identifiers},
+            )
+            if len(results) > 1
+            else Q(
+                "term",
+                **{"identifier.keyword": all_result_identifiers[0]},
             )
```
This is fine, but FWIW I don't think it's entirely necessary. `terms` should still work even with a list of 1 and I'd be surprised if there is a significant performance difference between the two. Have you already measured that? The only reason I'd suggest not doing this is (a) it feels a little like pre-optimising and (b) it makes the code slightly more obscure.

sarayourfriend commented on api/api/utils/search_context.py:
```diff
@@ -1,32 +1,35 @@
 from api.constants.media_types import OriginIndex
 
 
+class Identifiable(Protocol):
+    identifier: str
```
Cool!

sarayourfriend commented on api/api/views/media_views.py:
```diff
@@ -87,6 +89,20 @@ def get_db_results(self, results):
     # Standard actions
 
+    def retrieve(self, request, *_, **__):
+        @dataclass
+        class PseudoHit:  # implements Identifiable protocol
+            identifier: str
+
+        instance = self.get_object()
+        hit = PseudoHit(identifier=str(instance.identifier))
+
+        search_context = SearchContext.build([hit], self.default_index).asdict()
```
FWIW, `SearchContext::build` could be simplified if it just received the list of identifiers. Considering it gets mapped in the `::build` function out of the `results` list anyway, and those results aren't used in any _other_ way, I think it would be a nice way to simplify the `::build` function and the type juggling we're doing. It prevents us from needing to create a thin container class to satisfy `SearchContext` and reduces `SearchContext`'s concerns by restricting it to just the identifiers, the part of the results it actually cares about.

```suggestion
        instance = self.get_object()
        search_context = SearchContext.build([instance], self.default_index).asdict()
```

sarayourfriend reviewed (changes requested):
LGTM! My only request for a change is to add a unit or integration test against the single result endpoint to check that the `unstable__sensitivity` is populated when we expect it to be. We can use the local testing terms found in `mock_sensitive_terms.txt` to help find a result marked sensitive in the local data to test with.

The other suggestions to reduce `SearchContext` to just the identifiers is just a suggestion :slightly_smiling_face:

dhruvkb commented:
Added integration tests in 15c12b7 and simplified the `SearchContext` implementation to avoid dataclasses and protocols and deal with simple `list[str]` instead.

sarayourfriend reviewed (approved):
LGTM!

stacimc reviewed (approved):
Took me awhile to figure out how to set everything up for the manual testing, sorry 😅 Got everything working and this looks good to me!

dhruvkb commented:
Thank you for the effort @stacimc!
